File: bindings/java/src/main/java/org/indunet/robot/bus/TypedActionGoalHandle.java
package org.indunet.robot.bus;

import com.google.protobuf.MessageLite;
import java.util.concurrent.CompletableFuture;

/** Typed view of an in-flight action goal. */
public final class TypedActionGoalHandle<
                Feedback extends MessageLite, Result extends MessageLite>
        implements AutoCloseable {
    private final ActionGoalHandle inner;
    private final Class<Result> resultType;

    TypedActionGoalHandle(ActionGoalHandle inner, Class<Result> resultType) {
        this.inner = inner;
        this.resultType = resultType;
    }

    public String goalId() {
        return inner.goalId();
    }

    public String actionName() {
        return inner.actionName();
    }

    public Result result() {
        return decode(inner.result());
    }

    public Result result(double timeoutSecs) {
        return decode(inner.result(timeoutSecs));
    }

    public CompletableFuture<Result> resultAsync() {
        return inner.resultAsync().thenApply(this::decode);
    }

    public CompletableFuture<Result> resultAsync(double timeoutSecs) {
        return inner.resultAsync(timeoutSecs).thenApply(this::decode);
    }

    public void cancel() {
        inner.cancel();
    }

    @Override
    public void close() {
        inner.close();
    }

    private Result decode(ActionMessage message) {
        return ProtoCodec.parse(resultType, message.getBody());
    }
}

File: bindings/android/src/main/kotlin/org/indunet/robot/bus/TypedActionGoalHandle.kt
package org.indunet.robot.bus

import com.google.protobuf.MessageLite
import java.util.concurrent.CompletableFuture

/** Typed view of an in-flight action goal. */
class TypedActionGoalHandle<Feedback : MessageLite, Result : MessageLite> internal constructor(
    private val inner: ActionGoalHandle,
    private val resultType: Class<Result>,
) : AutoCloseable {
    fun goalId(): String = inner.goalId()

    fun actionName(): String = inner.actionName()

    @JvmOverloads
    fun result(timeoutSecs: Double = -1.0): Result =
        ProtoCodec.parse(resultType, inner.result(timeoutSecs).body)

    @JvmOverloads
    fun resultAsync(timeoutSecs: Double = -1.0): CompletableFuture<Result> =
        inner.resultAsync(timeoutSecs).thenApply { ProtoCodec.parse(resultType, it.body) }

    fun cancel() = inner.cancel()

    override fun close() = inner.close()
}
